src/github.com/getlantern/flashlight/client/handler.go

```diff
@@ -114,13 +114,11 @@ func (client *Client) intercept(resp http.ResponseWriter, req *http.Request, op
 // pipeData pipes data between the client and proxy connections.  It's also
 // responsible for responding to the initial CONNECT request with a 200 OK.
 func pipeData(clientConn net.Conn, connOut net.Conn, op *ops.Op, closeFunc func()) {
-	writeErrCh := make(chan error)
+	writeErrCh := make(chan error, 1)
 	// Start piping from client to proxy
 	op.Go(func() {
 		_, writeErr := io.Copy(connOut, clientConn)
-		if writeErr != nil {
-			writeErrCh <- writeErr
-		}
+		writeErrCh <- writeErr
```

Comment:
This was the main source of the leak.  In the case of successful copies, we never pushed anything onto writeErrCh, which would prevent the read down below, which prevented the call to close the connections.